src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,258 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in a graph.
+        /// This uses Tarjan's algorithm for finding strongly-connected components, or SCCs,
+        /// of the graph and processes those SCCs individually for finding cycles.
+        ///
+        /// This returns a list of cycles, each represented as a list of nodes. Nodes
+        /// for this algorithm are integers. The cycles are guaranteed to not contain
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// </summary>
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph)
+        {
+            // Possible Optimization: Each SCC could be processed in parallel.
+
+            // Possible Optimization: Getting the subgraph for an SCC only needs
+            // to consider the parent graph, not the full graph. A recursive strategy
+            // may be better than using a stack object.
+
+            var cycles = new List<List<int>>();
+
+            PushSCCs(graph);
+            while (SccStack.Any())
+            {
+                var (scc, startNode) = SccStack.Pop();
+                var subGraph = GetSubGraph(graph, scc);
+                cycles.AddRange(GetSccCycles(subGraph, startNode));
+
+                subGraph.Remove(startNode);
+                foreach (var (_, children) in subGraph)
+                {
+                    children.Remove(startNode);
+                }
+
+                PushSCCs(subGraph);
+            }
+
+            return cycles;
+        }
+
+        /// <summary>
+        /// Uses Tarjan's algorithm for finding strongly-connected components, or SCCs of a
+        /// graph to get all SCC, order them by their minimum-valued node, and push them to
+        /// the SCC stack.
+        /// </summary>
+        private void PushSCCs(Dictionary<int, List<int>> graph)
+        {
+            var sccs = TarjanSCC(graph).OrderByDescending(x => x.MinNode);
+            foreach (var scc in sccs)
+            {
+                SccStack.Push(scc);
+            }
+        }
+
+        private Dictionary<int, List<int>> GetSubGraph(Dictionary<int, List<int>> inputGraph, HashSet<int> subGraphNodes) =>
+            inputGraph
+                .Where(kvp => subGraphNodes.Contains(kvp.Key))
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Where(dep => subGraphNodes.Contains(dep)).ToList());
```

Comment:
Is it helpful to define an SCC as a hashset? If it was a list, we could traverse it linearly and have the same O(1) get from the inputGraph to create a new dictionary. Having it as a hashset is likely wasteful (most SCCs will contain a single element, anyway).

Reply:
Maybe I am misunderstanding your comment. I am currently under the impression that a HashSet and a List have similar size complexities, and that checking for an item in a List was O(n). If this is not the case, do you have any resources I could look at to better understand this?

Reply:
Dictionaries, sets, etc. usually have worse locality compared to array-based structures. I'd also expect them to have more memory overhead per structure. Basically, yes, it's O(1) for iterating over both but consts might be considerably different.

Reply:
I am not following. Let's discuss your suggestion over a call.

Reply:
Resolved offline: Because we need to do filtering on both the keys of the input graph, and the adjacency list values, it is easier to have the subgraph nodes structure be a hash set for doing `.Contains` operations.